public/tutorials/debugging-a-smart-contract.md

```diff
@@ -0,0 +1,621 @@
+# Debugging a smart contract
+
+<p class="alert alert-info">
+<strong>TODO</strong>: { [String: '4'] s: 1, e: 0, c: [ 4 ] } ----- What does all this stuff mean?
+</p>
+
+<p class="alert alert-info">
+<strong>TODO</strong>: How is the gas limit set? When/how does an out of gas error occur?
+</p>
+
+<p class="alert alert-info">
+<strong>TODO</strong>: Link to docs
+</p>
+
+
+## Background
+
+A smart contract in Ethereum is just code. Unlike the "paper" contracts that you find elsewhere, this contract needs to make sense in a very precise manner.
+
+(And that's a good thing. Imagine how much clearer real-world contracts would be if they needed to "compile"?)
+
+If our contracts are not coded correctly, our transactions may fail, which can result in the loss of ether (in the form of gas), not to mention wasted time and effort.
+
+**Luckily, Truffle (as of version 4) has a built in debugger for stepping through your code.** So when something goes wrong, you can find out exactly what it was, and fix it promptly.
+
+In this tutorial, we will migrate a basic contract to a test blockchain, introduce some errors into it, and solve each one through the use of the built-in Truffle debugger.
+
+<p class="alert alert-info">
+<strong>Note</strong>: This tutorial requires Truffle version 4.0 or newer.
+</p>
+
+
+## A basic smart contract
+
+One of the most basic, non-trivial, types of smart contract is a **simple storage contract**. (This example was adapted from the [Solidity documentation](https://solidity.readthedocs.io/en/develop/introduction-to-smart-contracts.html).)
+
+This contract does two things: 
+
+* Allows you to set a variable (`myVariable`) to a particular integer value
+* Allows you to query that variable to get the selected value
+
+```javascript
+pragma solidity ^0.4.0;
+
+contract SimpleStorage {
+  uint myVariable;
+
+  function set(uint x) {
+    myVariable = x;
+  }
+
+  function get() constant returns (uint) {
+    return myVariable;
+  }
+}
+```
+
+This isn't a very interesting contract, but that's not the point here. We want to see what happens when things go wrong.
+
+But first, let's set up our environment.
+
+## Deploying the basic smart contract
+
+1. Create a new directory where we will house our contract locally:
+
+   ```shell
+   mkdir simple-storage
+   cd simple-storage
+   ```
+
+1. Create a bare Truffle project:
+
+   ```shell
+   truffle init bare
+   ```
+
+1. Inside the `/contracts` directory, create a file called `Store.sol` and populate it with the following content:
```

Comment:
Minor nitpick: `contracts/` is a more common convention, I believe (for describing a local directory)

Reply:
Good by me.